### PR TITLE
Saa 74 refactoring the API model and DB model on the back of our understanding of the activties domain at this point in time

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
**Changes include:**

- Renaming model and entity ACTIVITY_SESSION ->  ACTIVITY_SCHEDULE
- Renaming model and entity ACTIVITY_INSTANCE -> SCHEDULED_INSTANCE
- Renaming model and entity ACTIVITY_PRISONER -> ALLOCATION
- Renaming model and entity ACTIVITY_WAITING -> PRISONER_WAITING
- Updating the DDL to reflect the names changes to entities and changing time attributes to actual time datatypes.
- Adding additional test data for iep level and pay bands for allocations in the integration test.

**Note: This contains breaking changes. It requires the DB to be dropped before deployment to the dev environment.**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivitySchedule.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+@Schema(
+  description =
+  """
+  Describes the weekly schedule for an activity. There can be several of these defined for one activity.
+  An activity schedule describes when, during the week, an activity will be run and where.
+  e.g. Tuesday PM and Thursday AM - suitable for Houseblock 2 to attend.
+  e.g. Monday AM and Thursday PM - suitable for Houseblock 3 to attend.
+  """
+)
+data class ActivitySchedule(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivitySchedule.kt:17]: Renamed from ActivitySession.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ScheduledInstance.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model
+
+import com.fasterxml.jackson.annotation.JsonFormat
+import io.swagger.v3.oas.annotations.media.Schema
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.LocalTime
+
+@Schema(description = "Describes one instance of an activity schedule")
+data class ScheduledInstance(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ScheduledInstance.kt:10]: Renamed from ActivityInstance.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivitySchedule.kt (hunk 1 above)
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivitySchedule.kt:17]: Renamed from ActivitySession.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivitySchedule.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  @Schema(description = "The description of this activity schedule", example = "Monday AM Houseblock 3")
+  val description: String,
+
+  @Schema(description = "If not null, it indicates the date until which this schedule is suspended", example = "10/09/2022")
+  @JsonFormat(pattern = "dd/MM/yyyy")
+  val suspendUntil: LocalDate? = null,
+
+  @Schema(description = "The time that any instances of this schedule will start", example = "9:00")
+  @JsonFormat(pattern = "HH:mm")
+  val startTime: LocalTime,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivitySchedule.kt:37]: Also include time changes to the model.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ScheduledInstance.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  @Schema(description = "The internally-generated ID for this scheduled instance", example = "123456")
+  val id: Long,
+
+  @Schema(description = "The specific date for this scheduled instance", example = "30/09/2022")
+  @JsonFormat(pattern = "dd/MM/yyyy")
+  val date: LocalDate,
+
+  @Schema(description = "The start time for this scheduled instance", example = "9:00")
+  @JsonFormat(pattern = "HH:mm")
+  val startTime: LocalTime,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ScheduledInstance.kt:21]: Time changes to the model here.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt [hunk 5]
```diff
@@ -16,23 +16,23 @@ import javax.persistence.OneToMany
...
-  val instances: MutableList<ActivityInstance> = mutableListOf(),
+  val instances: MutableList<ScheduledInstance> = mutableListOf(),
 
-  @OneToMany(mappedBy = "activitySession", fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
+  @OneToMany(mappedBy = "activitySchedule", fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
   @Fetch(FetchMode.SUBSELECT)
-  val prisoners: MutableList<ActivityPrisoner> = mutableListOf(),
+  val allocations: MutableList<Allocation> = mutableListOf(),
 
   val description: String,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:37]: Were we going to make the `startTime` and `endTime` into just time fields too? ie. they don't relate to any particular datetimes. Or a separate PR?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt [hunk 6]
```diff
@@ -11,13 +11,13 @@ data class ActivityPay(
   @Schema(description = "A list of pay bands and rates which apply to this activity pay. Can be empty if pay bands do not apply.")
   val bands: List<ActivityPayBand> = emptyList(),
 
-  @Schema(description = "The incentive/earned privilege basic rate per session for this activity", example = "100")
+  @Schema(description = "The incentive/earned privilege basic rate per schedule for this activity", example = "100")
   val iepBasicRate: Int? = null,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt:15]: These really are `sessions` - ie. the pay rate for one AM or PM session.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPayBand.kt [hunk 7]
```diff
@@ -11,7 +11,7 @@ data class ActivityPayBand(
   @Schema(description = "The pay band code - usually A-F - to differentiate different pay rates", example = "A")
   val payBand: String? = null,
 
-  @Schema(description = "The rate to be paid for one session of this activity", example = "220")
+  @Schema(description = "The rate to be paid for one schedule of this activity", example = "220")
   val rate: Int? = null,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPayBand.kt:15]: Same here - really are `sessions`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt (hunk 6 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt:15]: To be fair I did think this as well so perhaps should has left as is.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ScheduledInstance.kt (hunk 4 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ScheduledInstance.kt:21]: Should they come from the DB (as times) through the entity (as times)?

**Review by timharrison-moj — DISMISSED** [review]
LGTM. Just a question about whether times (not datetimes) should be in the DB and entities for schedules?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt (hunk 5 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:37]: I was kinda on the fence with this one.  I took the view just use the transformation layer for now but happy to take it all of the way to the DB ...

File: src/main/resources/migration/common/V1__baseline.sql [hunk 8]
```diff
@@ -114,77 +114,77 @@ CREATE TABLE activity_eligibility (
...
-CREATE TABLE activity_session (
-  activity_session_id           bigserial    NOT NULL CONSTRAINT activity_session_pk PRIMARY KEY,
+CREATE TABLE activity_schedule (
+  activity_schedule_id          bigserial    NOT NULL CONSTRAINT activity_schedule_id PRIMARY KEY,
   activity_id                   bigint       NOT NULL REFERENCES activity (activity_id),
   description                   varchar(50)  NOT NULL,
   suspend_until                 date,
-  start_time                    timestamp    NOT NULL,
-  end_time                      timestamp,
+  start_time                    time         NOT NULL,
```
Comment (matt-crowson-moj) [inline src/main/resources/migration/common/V1__baseline.sql:122]: Now times.

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**